### ISSUE TITLE
Does this package works with HMVC

### ISSUE THREAD

**Issue by Ihabafia** [Patroklo/codeigniter-static-laravel-routes#1]
Hi,

Is it possible to have this package to work as CI HMVC with the MX_Controller, instead of CI_Controller.

I really like this package so me.

Thanks for your effort.

**Comment by Patroklo** [comment]
Hi, I have tried this library with https://github.com/jenssegers/codeigniter-hmvc-modules this hmvc module and it worked just fine, but I had to change code there, because both libraries share the method "_parse_routes". You can just extend the hmvc's Route.php from my library instead of the usual CI_Route, then it should work just fine, because all code should be called apropriately. 

There is also the hmvc library from bitbucket, haven't tried that, but I don't see any shared method, so you can just extend my library from their Route.php instead of CI_Route and it should work just fine.

Don't know if I have expressed myself clearly enough, if not, just tell and I'll try again.

Thanks!

**Comment by Ihabafia** [comment]
Thank you for you reply,

I am using the MX HMVC from Bitbucket, but when I tried to install it, I
got the following error:

_Fatal error_: Access level to MY_Router::_set_routing() must be public (as
in class CI_Router) in _/Volumes/Data/newaum/application/core/MY_Router.php_
on line

_18_

_Could you please help._

_Thank you,_

_Ihab Abu Afia_

**Comment by Patroklo** [comment]
Ok, I'll check it. Which version of codeigniter are you using? Because _set_routing it's protected in my codeigniter installation...

**Comment by Ihabafia** [comment]
v.2.2

**Comment by Patroklo** [comment]
Ok, the problem is that there is no support for 2.X versions. Try changing the scope of the _set_routing method from my library from protected (that is the scope of 3.X versions) to public and it may work (but it's not sure, as i have said, I haven't checked this library for 2.X versions, but the most probable outcome it's that it wont work).

**Comment by Ihabafia** [comment]
I did tred that but i got more errors.
I hope if you can try implimenting it.

**Comment by Patroklo** [comment]
And updating your instalation is also not an option? I can't quite remember right now, but I think that it was not compatible with 2.X versions because the way the routes work there was different. I'll try to check it out.

**Comment by Patroklo** [comment]
Ok, just checked with a 2.2 version and with very little changes works just fine. This weekend I'll do a little manual with instructions and I'll try to upload a new branch with the changes for version 2.X.

**Comment by Ihabafia** [comment]
Great.  Thank you so much for your effort.

**Comment by Ihabafia** [comment]
I jist want to confirm that i am using bitbucket HMVC with CI2.2.

Thank you.

**Comment by Patroklo** [comment]
Hi, check this https://github.com/Patroklo/codeigniter-static-laravel-routes/tree/for-codeigniter-2.X and tell me if you have any issue. If the installation tutorial it's not clear enough, let me know (english it's not my first language).

**Comment by Patroklo** [comment]
Hi, have you been able to use the library with the new version?

**Comment by Ihabafia** [comment]
I am sorry.  I couldn't have time.  I was busy with another project.  I am
going to check it out in couple of days.

Thank you so much.

**Comment by Ihabafia** [comment]
I so sorry,

I just got time to use it, it's giving an error as follow:

Fatal error: Access level to MX_Router::_set_default_controller() must be public (as in class CI_Router) in /Users/ehabafia/Dropbox/WebProjects/ci2laravel/application/third_party/MX/Router.php on line 241

and line 241 is the last line in the class which have only 
241: }

**Comment by Ihabafia** [comment]
I think it's my fault, I was using HMVC for CI3, while I suppose to use HMVC for CI2.

Now it's working fine.

But the text that is in .gitignore in the routes folder is showing on the view:

# Ignore everything in this directory \* # Except this file !.gitignore

**Comment by Patroklo** [comment]
Well, it's a relief it all worked just fine! The .gitignore it's a known bug that I already repaired at the ci 3 version but didn't at 2, I'll check it right now.

**Comment by Patroklo** [comment]
Check if this change works if you can, please

**Comment by Ihabafia** [comment]
I noticed also that when you tried :+1: 

Route::resources('user');

it will give you an error since it's expecting $options to be array() while if you do it the way it is above, there will be no options at all.  The way I solved it is adding this on libraries/Route.php around line 614:

```
    if(!is_array($options)){
        $options = array();
    }
```

Then It worked just fine :)

**Comment by Ihabafia** [comment]
And I solved the other .gitignore problem as well as the .DS_Store for Mac as follows:
Change the line 29 in MY_Router.php to be:

```
if (is_file(APPPATH.'routes/'.$file) and $file != '.DS_Store' and $file != '.gitignore')
```

and it also worked perfect. :)

**Comment by Ihabafia** [comment]
But I couldn't understand when and how and why would I use this:

Route::head('user',         'user/index');
Route::patch('user/(:any),  'user/update/$1');
Route::options('user/(:any),'user/load/$1');

**Comment by Ihabafia** [comment]
Route filters are not working at all, I tried an example even the Anonymous function is not working as well: 

```
Route::any('user/{id}', 'user/load/$1', array('before' => function(){
    if(true)
    {
        echo "Filter Triggered"; exit;
    }
}));
```

it keep executing the load method on the user class/module

Any help?

**Comment by Ihabafia** [comment]
Please se why the Filter is not working.

Thank you,

**Comment by Ihabafia** [comment]
I tried the Filters on CI3 also but still not working. :(

**Comment by Ihabafia** [comment]
Also I tried the Route::resources('user'); is giving an error and to solve it, you still need to use: 

if(!is_array($options)){
$options = array();
}

I am using PHP v. 5.5.23

**Comment by Ihabafia** [comment]
I even removed the HMVC and run it as MVC out of the box for CI3, but the filters is still not working at all.

**Comment by Patroklo** [comment]
Let me check it.

**Comment by Ihabafia** [comment]
Frankly, I would say nothing is working except the basic route.

Even

Route::set_name('user_update', 'admin/user/load/');
Route::get('user', 'user/index', array('as' => 'user'));

are not working, except if I echo Route::named('user_update'); in the same route file, but if I use it in the controller or so, it forget the route name.

and it worked only that way with ::set_name but not array('as'=>'user'). neither with HMVC or even without.

Please Help.

**Comment by Patroklo** [comment]
Ok, I have just checked filters and they work just fine. So, let's check all your installation first:
The controller you are calling inherits My_Controller?
Have you added to Routes.php the Route::map() ?

**Comment by Patroklo** [comment]
Also the example of 
Route::any('user/{id}', 'user/load/$1', array('before' => function(){
    if(true)
    {
        echo "Filter Triggered"; exit;
    }
}));

works just fine. I think that the problem may be one or both of the things I have added in the previous post, I have changed my controller from inheriting MY_Controller to CI_Controller and stopped working (that's normal, all the filter callings are in that class).

**Comment by Patroklo** [comment]
About the usefulness of:

Route::head('user', 'user/index');
Route::patch('user/(:any), 'user/update/$1');
Route::options('user/(:any),'user/load/$1');

Well... they are useful for apis, for example. Usually nobody will use them, though.

**Comment by Ihabafia** [comment]
class User extends MX_Controller {

```
public function __construct() {
    parent::__construct();
}

public function show($id)
{
    echo Route::named('user_update');
    echo "Show User # ".$id;
}
```

}

**Comment by Ihabafia** [comment]
Because I am using HMVC

**Comment by Patroklo** [comment]
Check in the tutorial:

tl;dr install HMVC library, install the router library and make MY_Controller extend MX_Controller and MY_Router extend the HMVC Router library.

If you want filters or something more than only routing you have to use MY_Controller; that's because that class has a lot of logic neccessary to use that options.

I'm talking by memory here (haven't installed mvc libraries in my example) but you have to make my_controller extend mx_controller instead and use always my_controller

**Comment by Ihabafia** [comment]
Perfect the Filters is now working :+1: 

But the Name is not.

**Comment by Ihabafia** [comment]
Route::set_name('user_update', 'admin/user/load/');
Route::get('user', 'user/index', array('as' => 'user'));

are not working, except if I echo Route::named('user_update'); in the same route file, but if I use it in the controller or so, it forget the route name.

and it worked only that way with ::set_name but not array('as'=>'user'). neither with HMVC or even without.

Please Help.

**Comment by Ihabafia** [comment]
Route::set_name('user_update', 'admin/user/load/');
works fine when you echo it from the route file, but when you try to use it in the controller let's say, you get nothing.

but if you use the:
Route::get('user', 'user/index', array('as'=>'user'));
and you try to echo the Route::named('user'); it didn't work on the route file nor in the controller.

**Comment by Patroklo** [comment]
Mmmmmm they work for me... the my_router extends mx_router?

**Comment by Patroklo** [comment]
What error do you get with route::named? Or it sends a NULL?

**Comment by Ihabafia** [comment]
```
include_once(APPPATH.'core/Base_Router.php');

class MY_Router extends Base_Router {
```

**Comment by Ihabafia** [comment]
require APPPATH."third_party/MX/Router.php";

class Base_Router extends MX_Router {

**Comment by Patroklo** [comment]
im gonna try using hmvc, which version are you using of CI and hmvc?

**Comment by Ihabafia** [comment]
CI3 HMVC, any why he doesn't have HMVC for CI2 anymore :(

**Comment by Ihabafia** [comment]
wiredesignz Modular extensions - HMVC

**Comment by Patroklo** [comment]
ooook, so CI 3 and last version of hmvc, give me some time to download and try

**Comment by Ihabafia** [comment]
What error do you get with route::named? Or it sends a NULL?

I got nothing balck

**Comment by Ihabafia** [comment]
Thank you for your help, I am waiting.

**Comment by Patroklo** [comment]
Maaaan, it works for me XD Let's do this:

Go to libraries/route.php , method public static function named

Make a var dump at the next line of the method definition of self::$named_routes and let me know what do you get.

**Comment by Patroklo** [comment]
of course you should call Route::named('whatever'); at controller.

**Comment by Patroklo** [comment]
Also, where do you make a include_once of Route.php library? You may be calling it two times and it could be erasing the memory of the previous call?

**Comment by Ihabafia** [comment]
it doesn't seem to be reaching there, because it's not giving anything.  

Regarding the include_one, I am calling it once at /core/MY_Router.php

**Comment by Ihabafia** [comment]
```
include_once(APPPATH.'libraries/Route.php');
require APPPATH."third_party/MX/Router.php";

class MY_Router extends MX_Router {
```

**Comment by Patroklo** [comment]
Let's do this, I'm going to post a link with my installation, compare it to your installation for differences, because I'm sure it's a problem in some file where it's called wrong something and we can get crazy looking for it this way

**Comment by Patroklo** [comment]
http://www.filedropper.com/codeigniterapplication

**Comment by Patroklo** [comment]
I think the differences should be at core/ directory.

**Comment by Patroklo** [comment]
Of course, Im using last version of CI 3 and MX, only welcome controller and nothing more, a simple example. routes, i   "/prueba" calls for what we are looking for, don't know if you have changed the index.php/ thing.

**Comment by Ihabafia** [comment]
it worked perfect, god knows what was the difference, but I think you still have a problem when you do the following:

Route::resources('user');

if you user domain.dev/user or domain.com/user/5 or domain.com/user/5/edit you will get a big error, not sure if it will work in your example, but for sure it does work if you use it through the module.

I am going to test it by routing it to the flat controller.

**Comment by Patroklo** [comment]
Ok, so now everything it's working. Cool! I'll check the resources... tomorrow, I have a back hernia and it hurts too much not being stand down.

**Comment by Ihabafia** [comment]
Sorry to hear so,  Take care.

**Comment by Ihabafia** [comment]
Also, please help me to use the named dynamically in this example:

Route::get('user/{id}/edit', 'user/edit/$1',array('as'=>'edit'))->where('id', '[0-9]+');

In the controller I need to redirect the page by the route:

redirect(Route::named('edit'));

if the URL is /user/4/edit

when it gets the ::named('edit') it gives: user/$1/edit while it suppose to give user/4/edit.

Thank you,

**Comment by Patroklo** [comment]
Have you tried the example I have in the manual?

Optionally, if that route has uri parameters, you can set them via array at the second parameter of the method:

redirect(Route::named('user', array('12')));

You have to send the 4 via array because the system don't get automatically the data via uri given that usually user could want to add it manually.

**Comment by Ihabafia** [comment]
You didn't get me my friend, 

I am talking about a dynamic URI

if I use domain.dev/user/5/edit, and I made the route like this:

Route::get('user/{id}/edit', 'user/edit/$1',array('as'=>'edit'))->where('id', '[0-9]+');

then I need to redirect the page to that same page, so I need to type in the controller: 

redirect(Route::named('edit'));

so if I understood I suppose to make it like this:

redirect(Route::named('edit', array($id)); instead?

how about the error that I told you about in Route::resources('user');

Thank you so much.

**Comment by Patroklo** [comment]
yes, you have to write always the $1 in the named() method, oterwise, it will be used the string.

about the resources error, still haven't tried.

**Comment by Ihabafia** [comment]
ok,

as I told you my solution was add these to lines at the beggining of the method on the libraries/Route.php file on around line 612:

```
static function createRoute($from, $to, $options=array(), $nested=false)
{
    if(!is_array($options))
        $options = array();

    if(!is_null(self::$active_subdomain))
```

and it worked fine, but I am not sure if this is a good way.

**Comment by Patroklo** [comment]
I have made a change at the file, it should work just fine now, thanks for the info!

**Comment by Ihabafia** [comment]
Great work, and a perfect library.

I wish you the best.

**Comment by Ihabafia** [comment]
I can't access the css and js assets folder because of the routing problem, how could I solve that?

domain.dev/_assets/css/common/select2.min.css

when I access it through the browser it's giving file not fount.

which they are on the folder /application/_assets/css/common/select2.min.css

Thank you,
